ruby/examples/lib/apply_order_discount.rb

```diff
@@ -0,0 +1,44 @@
+class ApplyOrderDiscount
+  Request = Struct.new(:id)
+  Response = Struct.new(:discount, :has_discount?)
+
+  attr_reader :gateway
+
+  def initialize(gateway: Gateway.new)
+    @gateway = gateway
+  end
+
+  def apply(request)
+    order = gateway.find_order_by_id(request.id)
+
+    unless order.has_discount?
+      order.discount = discount_amount
+      gateway.save_order_discount(request.id, order)
+    end
+
+    Response.new(order.discount, order.has_discount?)
```

Comment:
Interesting that the response returns the discount. I wonder what the users of this apply use case really need? 